snapshotter/devmapper/devmapper.go

```diff
@@ -312,18 +310,44 @@ func (dm *Snapshotter) buildMounts(snap storage.Snapshot) []mount.Mount {
 	return mounts
 }
 
-func complete(ctx context.Context, trans storage.Transactor, err error) error {
-	if err == nil {
+// withTransaction wraps fn callback with containerd's meta store transaction.
+// If callback returns an error or transaction is not writable, database transaction will be discarded.
+func (dm *Snapshotter) withTransaction(ctx context.Context, writable bool, fn func(ctx context.Context) error) error {
+	ctx, trans, err := dm.store.TransactionContext(ctx, writable)
```

Comment:
It seems `writable` parameter isn't needed here. Why not add a new helper function instead?
instead of
```go
	return dm.withTransaction(ctx, false, func(ctx context.Context) error {
		return storage.WalkInfo(ctx, fn)
	})
```
how about
```go
// probably needs a better name
func threadSafeCall(ctx context.Context, fn func(ctx context.Context) error) error  {
	m := sync.Mutex
	m.Lock()
	defer m.Unlock()
	return fn(ctx)
}
```
```go
	return threadSafeCall(ctx, func(ctx context.Context) error {
		return storage.WalkInfo(ctx, fn)
	})
```

Reply:
Transactions are required by containerd's `storage` package, even if they are readonly.  Also see comment [here](https://github.com/containerd/containerd/blob/adfaa697a8b778eb319a49c02406c84bc0f1bdf9/snapshots/storage/metastore.go#L41-L44).

Reply:
Interesting. Alright, lgtm